### PR TITLE
Fix compilation with clang-cl on windows

### PR DESCRIPTION
**Pull request by DenizThatMenace**
When compiling with Visual Studio using `clang-cl` compiler the resource-compiler of MSVC will still be used. However, it trips over some clang-specific things in the C standard headers of `libc++` which come with `clang-cl`.

With the changes introduced in this PR the resource-compiler will no longer see these problematic headers etc.

### CODE REVIEW

**Comment by nmoinvaz** [comment]
I have not had any problems compiling with `cmake . -T ClangCl`. Is that what you are using?

**Comment by mtl1979** [comment]
It would be useful to know what version of clang-cl was used when the error occurred and what version the resource compiler was.

I've seen weird things happen when incompatible versions are used.

Like @nmoinvaz already said, the full command-line to invoke either `cmake` or `configure` helps us to reproduce the issue.

**Comment by DenizThatMenace** [comment]
One of my failing configurations is for example *Visual Studio 2019* in version **16.11.8** in *"x64 Native Tools Command Prompt for VS 2019"*:

```
C:\test_zlib-ng\build>where rc
C:\Program Files (x86)\Windows Kits\10\bin\10.0.20348.0\x64\rc.exe

C:\test_zlib-ng\build>where clang-cl
C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\Llvm\x64\bin\clang-cl.exe

C:\test_zlib-ng\build> where cmake
C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe

C:\test_zlib-ng\build> cmake --version
cmake version 3.20.21032501-MSVC_2
```
The output of cmaking and building is the following. (Note, I shortened the build-output but left the errors and some interesting warnings, that demonstrate some of the problems.)
```
C:\test_zlib-ng\build> cmake -G "Visual Studio 16 2019" -T ClangCL ..\source
-- Using CMake version 3.20.21032501-MSVC_2
-- ZLIB_HEADER_VERSION: 1.2.11
-- ZLIBNG_HEADER_VERSION: 2.1.0.devel
-- Selecting Windows SDK version 10.0.20348.0 to target Windows 10.0.22000.
-- The C compiler identification is Clang 12.0.0 with MSVC-like command-line
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/Llvm/x64/bin/clang-cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Arch detected: 'x86_64'
-- Basearch of 'x86_64' has been detected as: 'x86'
-- Architecture supports unaligned reads
-- Architecture supports unaligned reads of > 4 bytes
-- Looking for sys/sdt.h
-- Looking for sys/sdt.h - not found
-- Looking for unistd.h
-- Looking for unistd.h - not found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of off64_t
-- Check size of off64_t - failed
-- Check size of _off64_t
-- Check size of _off64_t - failed
-- Check size of __off64_t
-- Check size of __off64_t - failed
-- Looking for fseeko
-- Looking for fseeko - not found
-- Looking for strerror
-- Looking for strerror - found
-- Looking for posix_memalign
-- Looking for posix_memalign - not found
-- Performing Test HAVE_NO_INTERPOSITION
-- Performing Test HAVE_NO_INTERPOSITION - Failed
-- Performing Test HAVE_ATTRIBUTE_VISIBILITY_HIDDEN
-- Performing Test HAVE_ATTRIBUTE_VISIBILITY_HIDDEN - Success
-- Performing Test HAVE_ATTRIBUTE_VISIBILITY_INTERNAL
-- Performing Test HAVE_ATTRIBUTE_VISIBILITY_INTERNAL - Success
-- Performing Test HAVE_BUILTIN_CTZ
-- Performing Test HAVE_BUILTIN_CTZ - Success
-- Performing Test HAVE_BUILTIN_CTZLL
-- Performing Test HAVE_BUILTIN_CTZLL - Success
-- Performing Test HAVE_PTRDIFF_T
-- Performing Test HAVE_PTRDIFF_T - Success
-- Performing Test HAVE_AVX2_INTRIN
-- Performing Test HAVE_AVX2_INTRIN - Success
-- Performing Test HAVE_SSE42CRC_INLINE_ASM
-- Performing Test HAVE_SSE42CRC_INLINE_ASM - Success
-- Performing Test HAVE_SSE42CRC_INTRIN
-- Performing Test HAVE_SSE42CRC_INTRIN - Failed
-- Performing Test HAVE_SSE42CMPSTR_INTRIN
-- Performing Test HAVE_SSE42CMPSTR_INTRIN - Failed
-- Performing Test HAVE_SSE2_INTRIN
-- Performing Test HAVE_SSE2_INTRIN - Success
-- Performing Test HAVE_SSSE3_INTRIN
-- Performing Test HAVE_SSSE3_INTRIN - Success
-- Performing Test HAVE_PCLMULQDQ_INTRIN
-- Performing Test HAVE_PCLMULQDQ_INTRIN - Success
-- Architecture-specific source files: arch/x86/x86.c;arch/x86/slide_hash_avx.c;arch/x86/chunkset_avx.c;arch/x86/compare258_avx.c;arch/x86/adler32_avx.c;arch/x86/insert_string_sse.c;arch/x86/chunkset_sse.c;arch/x86/slide_hash_sse.c;arch/x86/adler32_ssse3.c;arch/x86/crc32_fold_pclmulqdq.c
-- The following features have been enabled:

 * CMAKE_BUILD_TYPE, Build type: Release (default)
 * AVX2_SLIDEHASH, Support AVX2 optimized slide_hash, using "-mavx2"
 * AVX_CHUNKSET, Support AVX optimized chunkset, using "-mavx2"
 * AVX2_COMPARE258, Support AVX2 optimized compare258, using "-mavx2"
 * AVX2_ADLER32, Support AVX2-accelerated adler32, using "-mavx2"
 * SSE42_CRC, Support SSE4.2 optimized CRC hash generation, using "-msse4"
 * SSSE3_ADLER32, Support SSSE3-accelerated adler32, using "-mssse3"
 * PCLMUL_CRC, Support CRC hash generation using PCLMULQDQ, using "-mssse3 -msse4 -mpclmul"
 * WITH_GZFILEOP, Compile with support for gzFile related functions
 * ZLIB_ENABLE_TESTS, Build test binaries
 * WITH_SANITIZER, Enable sanitizer support
 * WITH_OPTIM, Build with optimisation
 * WITH_NEW_STRATEGIES, Use new strategies
 * WITH_UNALIGNED, Support unaligned reads on platforms that support it
 * WITH_UNALIGNED64, Support unaligned 64-bit reads on platforms that support it
 * WITH_AVX2, Build with AVX2
 * WITH_SSE2, Build with SSE2
 * WITH_SSSE3, Build with SSSE3
 * WITH_SSE4, Build with SSE4
 * WITH_PCLMULQDQ, Build with PCLMULQDQ

-- The following features have been disabled:

 * ZLIB_SYMBOL_PREFIX, Publicly exported symbols DO NOT have a custom prefix
 * ZLIB_COMPAT, Compile with zlib compatible API
 * ZLIB_DUAL_LINK, Dual link tests against system zlib
 * WITH_FUZZERS, Build test/fuzz
 * WITH_NATIVE_INSTRUCTIONS, Instruct the compiler to use the full instruction set on this host (gcc/clang -march=native)
 * WITH_MAINTAINER_WARNINGS, Build with project maintainer warnings
 * WITH_CODE_COVERAGE, Enable code coverage reporting
 * WITH_INFLATE_STRICT, Build with strict inflate distance checking
 * WITH_INFLATE_ALLOW_INVALID_DIST, Build with zero fill for inflate invalid distances
 * INSTALL_UTILS, Copy minigzip and minideflate during install

-- Configuring done
-- Generating done
-- Build files have been written to: C:/test_zlib-ng/build

C:\test_zlib-ng\build> cmake --build . --config Debug
Microsoft (R)-Build-Engine, Version 16.11.2+f32259642 für .NET Framework
Copyright (C) Microsoft Corporation. Alle Rechte vorbehalten.

  Checking Build System

...

  In file included from C:\test_zlib-ng\source\adler32.c:8:
  In file included from C:\test_zlib-ng\source/functable.h:9:
C:\test_zlib-ng\source/deflate.h(284,3): warning : attribute 'align' is ignored, place it after "struct" to apply attribute to type declaration [-Wignored-attributes] [C:\test_zlib-ng\build\zlib.vcxproj]
C:\\test_zlib-ng\source/zutil.h(257,34): message : expanded from macro 'ALIGNED_' [C:\test_zlib-ng\build\zlib.vcxproj]
...

clang-cl : warning : unknown argument ignored in clang-cl '-msse4'; did you mean '-msse2'? [-Wunknown-argument] [C:\test_zlib-ng\build\zlib.vcxproj]
C:\test_zlib-ng\source\arch\x86\crc32_fold_pclmulqdq.c(443,30): warning : implicit declaration of function '_mm_extract_epi32' is invalid in C99 [-Wimplicit-function-declaration] [C:\test_zlib-ng\build\zlib.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\Llvm\x64\lib\clang\12.0.0\include\stdint.h(20): warning RC4067: unexpected characters following '#if/#elif' directive; newline expected [C:\test_zlib-ng\build\zlib.vcxproj]

C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\Llvm\x64\lib\clang\12.0.0\include\stdint.h(20): warning RC4067: unexpected characters following '#if/#elif' directive; newline expected [C:\test_zlib-ng\build\zlib.vcxproj]

C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\Llvm\x64\lib\clang\12.0.0\include\vadefs.h(12): fatal error RC1021: invalid preprocessor command 'include_next' [C:\test_zlib-ng\build\zlib.vcxproj]

...

C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\Llvm\x64\lib\clang\12.0.0\include\stdint.h(20): warning RC4067: unexpected characters following '#if/#elif' directive; newline expected [C:\test_zlib-ng\build\zlibstatic.vcxproj]

C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\Llvm\x64\lib\clang\12.0.0\include\stdint.h(20): warning RC4067: unexpected characters following '#if/#elif' directive; newline expected [C:\test_zlib-ng\build\zlibstatic.vcxproj]

C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\Llvm\x64\lib\clang\12.0.0\include\vadefs.h(12): fatal error RC1021: invalid preprocessor command 'include_next' [C:\test_zlib-ng\build\zlibstatic.vcxproj]
```

**Comment by nmoinvaz** [comment]
I fixed some of the warnings you were encountering in my PR.

**Comment by nmoinvaz** [comment]
It seems this is due to the fact that we `#include "zlib.h"` inside `win32/zlib1.rc`.

**Comment by nmoinvaz** [comment]
To me it seems like a bug in Windows SDK. I don't think older versions of the SDK have this problem. Or older versions of ClangCl.

**Comment by nmoinvaz** [comment]
I have submitted the issue to Microsoft.
https://developercommunity.visualstudio.com/t/ClangCl-include_next-error-when-includi/1622642?space=62&entry=myfeedback

**Comment by exoosh** [comment]
@nmoinvaz the issue is caused by the fact that the Clang component offered with MSVC comes with some own "standard headers" (mostly compiler intrinsics, but also stuff like `stddef.h`, `stdarg.h` or `stdint.h`) which don't guard against the `RC_INVOKED` case in the same fashion as those from MSVC proper and the Windows SDK. See `%ProgramFiles(x86)%\Microsoft Visual Studio\2019\%EDITION%\VC\Tools\Llvm\lib\clang\%CLANGVER%\include` for details. (`EDITION=Professional` or `Enterprise` or `Community` [?], as of VS2019 `CLANGVER=12.0.0`). The result is a mix of LLVM-provided headers and those from the Windows SDK being handed to the resource compiler (`rc.exe`).

The usual preprocessor code to avoid this is along the lines of (gleaned from the Windows SDK):

```
#if !defined RC_INVOKED && !defined __midl
// ... part to be hidden from rc.exe and midl.exe ...
#endif
```

The ultimate cause is the rather limited preprocessor code employed by `rc.exe`. For this purpose a trimmed down `winnt.rh` (and other `.rh` files - I imagine it stands for "resource header") exist in the Windows SDK for inclusion inside `.rc` scripts. But by including `zlib.h`, which in turn included even more "standard" stuff, the resource compiler got to see headers it can't handle.

So it's not the Windows SDK that's the culprit, it's more of a mixed bag. Developers should generally guard against it (`RC_INVOKED` check, introduced by this PR), but also the (Windows) flavor of the LLVM/Clang headers that optionally comes with MSVC should probably fixed to hide unnecessary stuff from the resource compiler.

Edit 1: in a sense the Windows SDK could be blamed in part, too, if it recently changed which "standard headers" it includes from some SDK header. But ultimately the resource script asked for `zlib.h` and _its_ dependencies to be included and so should proactively shield `rc.exe` from any C++ preprocessor madness.

**Comment by nmoinvaz** [comment]
Because it effects not only zlib-ng but also zlib (which rarely gets updates and will probably never see the associated PR merged) I lean more toward it being fixed on the ClangCl side of things. Otherwise I would suggest we change code.

**Comment by exoosh** [comment]
> Because it effects not only zlib-ng but also zlib (which rarely gets updates and will probably never see the associated PR merged) I lean more toward it being fixed on the ClangCl side of things. Otherwise I would suggest we change code.

@nmoinvaz You're certainly right that this was probably a case of oversight on part of the VS team at Microsoft.

However, I think that the underlying issue is with zlib-ng and zlib respectively.

By providing a resource script, you have to abide by the rules of the resource compiler (`rc.exe`). The sole reason this worked in the past, despite inclusion of headers not meant for `rc.exe`, is that all the other headers included played nice and took care of checking for `RC_INVOKED`. It's unfortunate that the LLVM/Clang headers aren't prepared in the same fashion, but I think that a resource script should pay some attention to what gets included. It should avoid inclusion of unnecessary parts, if only for performance reasons and not for correctness.

So from my perspective the right thing would still be to apply this patch _here_ (even though this may not ultimately help fixing it in zlib).

That said, I agree that it should probably _also_ be addressed by the VS team such that the experience of switching between `cl` and `clang-cl` toolchains is as seamless as can be.

**Comment by nmoinvaz** [comment]
I agree with some of what you are saying, but I still think that it is on Microsoft to resolve the issue.

As for this patch:

- It may also be possible to move the version #define macros into a separate header file. So I'm not sure what the right approach would be. @Dead2 might have some thoughts on that.
- The zlib compat header also needs to be modified. 
- I'm not sure I like the comments use the word "harmful". It makes it seem like a serious issue maybe even security related, but it is not.

**Comment by exoosh** [comment]
> I agree with some of what you are saying, but I still think that it is on Microsoft to resolve the issue.

Full ack. For their part, it certainly is (LLVM/Clang as part of MSVC). I think this will eventually crop up in other projects, too. That alone should provide enough impetus for Microsoft to fix it 😁

> As for this patch:
> * It may also be possible to move the version #define macros into a separate header file. So I'm not sure what the right approach would be.

In my private projects (this account is work-related) this is exactly the pattern I follow, too. A separate header containing only the parts that both `rc.exe` and `cl.exe` are able to consume. This also has the benefit that the version info can be easily accessed from code (although on Windows I prefer reading it straight from the resources).

> * I'm not sure I like the comments use the word "harmful". It makes it seem like a serious issue maybe even security related, but it is not.

~I'm not following. Is this a comment from me you're referring to, @nmoinvaz? Searching for the term in this PR it only comes up in your comment.~ Dang, just looked at the diff. You're right, this is a bit strongly worded. Perhaps something @Bagira80 is willing to address.

**Comment by DenizThatMenace** [comment]
> * It may also be possible to move the version #define macros into a separate header file. So I'm not sure what the right approach would be. @Dead2 might have some thoughts on that.
> * The zlib compat header also needs to be modified.
> * I'm not sure I like the comments use the word "harmful". It makes it seem like a serious issue maybe even security related, but it is not.

You should be able to edit this PR to your liking or to provide another solution and just close it as obsoleted by that other solution. (I am passionless here.)

I would probably prefer the approach with a separate version header but I did not want to provide such a fix which changes too much (or diverges too much from the code of the original `zlib` library). I went for the least invasive solution.

**Comment by mtl1979** [comment]
Microsoft already use zlib as one of the QA tests for their own compiler, but it's quite obvious they don't use it for ClangCL yet.

Personally I think this falls in the category of bad compiler releases, that already include original releases of Visual C++ 2017 and 2019.

Nevertheless, if the workaround is something that doesn't break build on other supported compiler releases, then it might be worth merging.

**Comment by Dead2** [comment]
Regarding splitting version headers out, I am not sure whether that is a good idea or not.
I mean we could, but I'd like to keep the number of distributed header files low. It could perhaps go into zconf.h.

**Comment by mtl1979** [comment]
I don't think splitting out version numbers as own header is a good thing either... Any "workaround" we do should be minimal as eventually system headers in ClangCL must be fixed, or more and more projects will have issues...

**Comment by nmoinvaz** [comment]
With the latest `develop` branch you can pass in `-DBUILD_SHARED_LIBS=OFF` in cmake to skip building DLL sources and bypassing the error. But of course if you are actually trying to build the DLL, then you have the same error as before.

**Comment by exoosh** [comment]
> [...] as eventually system headers in ClangCL must be fixed, or more and more projects will have issues...

@mtl1979 While I agree that it _should also be fixed_ on their end in order to pass those headers to `rc.exe`, I think a fix inside _this project_ is a matter of correctness, since you pass the header for inclusion to `rc.exe`. If you have no control over these headers ([and their precedence](https://developercommunity.visualstudio.com/t/ClangCl-include_next-error-when-includi/1622642#T-N1627145)) you should simply hide them from `rc.exe`, especially since only [these seven defines](https://github.com/zlib-ng/zlib-ng/blob/b2ef108051610dc435939ebabaedf36ffd053984/zlib-ng.h.in#L48-L54) are _actually required_ by the resource script.

You would hardly expect a C compiler frontend to handle C++ code either. It's similar here, the `rc.exe` contains no full-fledged C++ preprocessor (I doubt it's capabilities are even fully on par with a C99 preprocessor). The choice to pass this to `rc.exe` must have been a conscious one by the author(s) of the resource script. Alone the fact that the headers that ship with MSVC proper and the Windows SDKs guard against inclusion when `RC_INVOKED`, has kept this from falling apart. I don't know if `windres.exe` (Binutils) and `llvm-rc.exe` (LLVM) define `RC_INVOKED`. But if they're compatible they probably do.

**Comment by mtl1979** [comment]
@exoosh This pull request hasn't been rejected yet, even though it's basically incomplete... Nobody has requested me to review it, so nobody can also claim that I rejected it on any grounds

Controversial patches need to be approved by two long-time contributors and currently there is 3 still active -- including me, @nmoinvaz and of course @Dead2.

**Comment by exoosh** [comment]
@mtl1979 I understand (your relationship to the project as Collaborator is visible on all of your comments; Member for the other two participants in this discussion). I merely meant to argue in favor of a fix within this project.

**Comment by mtl1979** [comment]
@exoosh I was just saying that when/if the original author of the pull request completes the fix, then it is likely that this pull request gets merged in...

**Comment by DenizThatMenace** [comment]
> @exoosh This pull request hasn't been rejected yet, even though it's basically incomplete... 
> ...
> @exoosh I was just saying that when/if the original author of the pull request completes the fix, then it is likely that this pull request gets merged in...

What is missing? Why is it incomplete? Because you do not want to modify it and you want me to remove the word "harmful" from the comments!?

**Comment by DenizThatMenace** [comment]
> What is missing? Why is it incomplete? Because you do not want to modify it and you want me to remove the word "harmful" from the comments!?

OK, I think I found out what you meant was missing. The same changes to `zlib.h.in`.

I added them and rebased my commit on the top of the current `develop` branch.

File: zlib-ng.h.in
Comment (nmoinvaz) [inline zlib-ng.h.in]: Perhaps it would read better if it said something along the lines of, "Only expose version information for the VS resource compiler."

File: zlib-ng.h.in
Comment (nmoinvaz) [inline zlib-ng.h.in]: Perhaps it would read better if it said something along the lines of, "Only expose version information for the VS resource compiler."

File: zlib-ng.h.in
Comment (nmoinvaz) [inline zlib-ng.h.in]: No comment needed here.

**Review by nmoinvaz — CHANGES_REQUESTED** [review]

**Comment by nmoinvaz** [comment]
> Because you do not want to modify it and you want me to remove the word "harmful" from the comments!?

That was me requesting that, not @mtl1979. I don't have the ability to edit the commit and I don't think @mtl1979 does either. I think the only one who does is @Dead2. Don't get frustrated if the PR is not accepted at first. I have had many PRs in the past remain open for several months in this repository and it has taken me many edits and rebases for some of them.

**Comment by mtl1979** [comment]
Even though most of us work on zlib-ng on our spare time, it doesn't mean we want the source code look like elementary school project with a lot of comments that suggest the developers have no clue that other people might read what they write and such write every comment with absolute no self-moderation. Most inline comments should be written in neutral tone and in third person perspective.

**Comment by DenizThatMenace** [comment]
> Even though most of us work on zlib-ng on our spare time, it doesn't mean we want the source code look like elementary school project with a lot of comments that suggest the developers have no clue that other people might read what they write and such write every comment with absolute no self-moderation. Most inline comments should be written in neutral tone and in third person perspective.

I am not taking this comment personally or as an insult (although some people might think it sounds like it). I just want you to note that I originally wrote the word "harmful" in the comment because I consider code that leads to compilation-errors as "harmful". (It harms the compilation process.) Maybe it is not the most precise word (I am not a native English speaker) but that is what I meant it to express. Nothing more, nothing less.

> I don't have the ability to edit the commit and I don't think @mtl1979 does either. I think the only one who does is @Dead2.

I checked the "Allow edits by maintainers" checkbox in order to allow those of you reviewing this PR to make small adjustments yourself if needed. (You will probably agree that this PR really contains very simple changes, so everyone should probably be able to adjust it without breaking it.) I did not expect this checkbox to only apply to a single person.  
I originally overlooked that `zlib.h.in` needed fixing, too. That's true. But your comment "The zlib compat header also needs to be modified." was not really helping me at first, seeing what header I need to adjust as well.

> Don't get frustrated if the PR is not accepted at first. I have had many PRs in the past remain open for several months in this repository and it has taken me many edits and rebases for some of them.

I am not frustrated if a PR takes several iterations with a lot of changes. (I already know this from several large and complex PRs I contributed to the `CMake` project.)  
I am also not frustrated if this PR does not get accepted. As I already said, I would be fine if this problem would get fixed by other means.  
I only would have been frustrated if you had considered this not to be a problem worth fixing.

I mean, it would not hurt me. I have a fix for it and I could maintain it myself and always apply it if I am compiling `zlib-ng` with Clang-Cl. But I think we can all agree that it would be better for `zlib-ng` to no longer have this problem and to integrate a fix for it into `zlib-ng` directly. Besides, I wanted to give something back to the community / the developers / you.

**Comment by mtl1979** [comment]
@Bagira80 We're all different, that's why it's good thing that more than one person reviews any commit... Eventually there will be more than just 3 persons who can approve "controversial" commits. Simple "non-controversial" commits only need one long-time contributor to approve.

Personally I don't expect every contributor to be 100% fluent in English, but we still try to avoid having too many spelling or grammar issues in the inline comments. We still have to compete with other zlib alternatives that might have more organized development team.

**Comment by DenizThatMenace** [comment]
@nmoinvaz I am not sure if you have noticed, but I applied your suggestions.

@mtl1979, @Dead2, @nmoinvaz:  
I consider this PR as being in its final form. (If not, please tell me what is still missing.)  
So it is now up to you to decide if it should be merged in its current form or not.

Regarding the failing Codecov test, I have no clue what it means in this context.  
The `RC_INVOKED` define conditionally excludes some code-parts which now no longer will be seen by the resource compiler. (And they should not.)  
Where the other differences come from I do not know.

**Comment by mtl1979** [comment]
I have nothing against merging except there should be no cancelled CI checks and as few failed checks as possible (code coverage checks are allowed to fail; fails fixed by other pending pull requests are also allowed, but affect merging order). 

We've had workarounds before and most likely will also need some more in the future when for example compiler defaults are changed to incompatible -- but more modern -- ones.

**Comment by DenizThatMenace** [comment]
@mtl1979

I do not know why some test are canceled. And I also do not know how the tests got triggered in the first place. (At least I haven't done so explicitly.) To be honest, I do not even know how to trigger the tests at all. (I assume a new commit might trigger these automatically... but then there seems to be a substantially long time between the commit and the test-run. In this case it seemed to be five days.)

Maybe some of you should re-trigger all tests explicitly and we will see how far they get?

**Comment by mtl1979** [comment]
> @mtl1979
> 
> I do not know why some test are canceled. And I also do not know how the tests got triggered in the first place. (At least I haven't done so explicitly.) To be honest, I do not even know how to trigger the tests at all. (I assume a new commit might trigger these automatically... but then there seems to be a substantially long time between the commit and the test-run. In this case it seemed to be five days.)
> 
> Maybe some of you should re-trigger all tests explicitly and we will see how far they get?

I don't have permission to trigger or re-trigger tests... I know @Dead2 can but not sure who else...

**Comment by Dead2** [comment]
All cancelled tests re-started. I have no idea why they were cancelled, probably a Github Actions failure/timeout.

For new contributors, I have to authorize the tests before it will start, so there was a delay for that.

In the future, an easy way to re-trigger all tests is to do a force-push (only needs an extra space in the commit message for example). I'll still need to re-authorize the CI run for new contributors, but that is one button instead of going into every single workflow and doing a re-start for each of them. (Heh, wonder whether there are helper-utils that do this for you)

**Comment by DenizThatMenace** [comment]
The check results looked better at first. (Almost 100% success rate.)  
Then I looked into the details of the single failing check (looks unrelated to my changes) and when I returned to the PR suddenly there were several "canceled" checks.

Coincidence? Or causality?

**Comment by DenizThatMenace** [comment]
And now suddenly (almost) 100% success rate again!? Maybe it was just a caching/render problem in my browser?

**Comment by Dead2** [comment]
pigz no-threads failure is already fixed in develop, so we can ignore that (or it would be fixed by a rebase).

**Comment by Dead2** [comment]
I am just wondering about one thing here, is it actually necessary to exclude the second block of code or just the include files?
It just seems to me that it is adding unnecessary complexity if the include files is where the errors come from..
That would also mean the first block would need to be smaller and only cover the included files, but IMO that would be preferable style-wise, if it is still a valid workaround for the problem.

**Comment by DenizThatMenace** [comment]
> I am just wondering about one thing here, is it actually necessary to exclude the second block of code or just the include files?

I tested it and the second blocks seems to not be strictly necessary.  

> It just seems to me that it is adding unnecessary complexity if the include files is where the errors come from.. That would also mean the first block would need to be smaller and only cover the included files, but IMO that would be preferable style-wise, if it is still a valid workaround for the problem.

That is not entirely going to work. The first block still has to cover the pre-processor block which checks that `ZCONFNG_H` (respective `ZCONF_H`) is defined because otherwise the `#error` directive would be triggered which would fail building with the resource compiler.

---

In general, I still think that conditionally hiding everything from the resource compiler except for the version defines is the correct way. (Ultimately, having a dedicated header only containing these version defines would be the best.)

**Comment by Dead2** [comment]
@Bagira80 Would you please change this to only block out a minimal amount of code? Then we can get this merged.
We'll have to consider doing a separate header file later.

**Comment by DenizThatMenace** [comment]
> @Bagira80 Would you please change this to only block out a minimal amount of code? Then we can get this merged. We'll have to consider doing a separate header file later.

Done.

BTW: I think you should have been able to adjust this PR yourself. (That's why I checked the "Allow edits by maintainers" checkbox.) Or do you also do not have write access to this PR?

**Comment by Dead2** [comment]
@Bagira80 Thanks 👍 
Yes, I have access to modify PRs, but I would rather avoid doing so, especially if I am unable to actually test the changes myself 😄 

I will merge this when it passes CI.

**Comment by Dead2** [comment]
Hmm, whatever happened to the Nmake CI? I don't remember any recent changes to that, so maybe Github Actions changed, again..
```
Run call "C:\Program Files (x8[6](https://github.com/zlib-ng/zlib-ng/runs/5095292083?check_suite_focus=true#step:3:6))\Microsoft Visual Studio\2019\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" x[8](https://github.com/zlib-ng/zlib-ng/runs/5095292083?check_suite_focus=true#step:3:8)6
The system cannot find the path specified.
'nmake' is not recognized as an internal or external command,
operable program or batch file.
Error: Process completed with exit code 1.
```

**Comment by mtl1979** [comment]
@Dead2 They upgraded to Visual Studio Enterprise 2022...

**Comment by exoosh** [comment]
> @Dead2 They upgraded to Visual Studio Enterprise 2022...

The appropriate method is to ask `vswhere` where to find those things; and whose path is _hardcoded_ to be at `%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe` on any machine with VS2002 through 2022.

I have [a script](https://github.com/assarbad/scripts/blob/master/setvcvars.cmd) among my private GitHub projects, which can detect an array of VS versions starting at VS2002 or so. Feel free to use it. By calling it without any additional parameters you will automatically get the latest installed VS version.

**Comment by mtl1979** [comment]
@exhoosh We might not always want the latest version available... Sometimes it's more useful to ask for specific version and let GitHub select correct runner supporting that version...

**Comment by exoosh** [comment]
That's equally supported (`setvcvars.cmd 2019` for example) and the script will even provide meaningful error output (and a non-zero exit code) when the requested version can't be found. I just meant to say that this is the easiest way of getting started with the script. But you needn't use it. I just wanted to offer it, as it is a convenient solution to the problem at hand.

**Comment by mtl1979** [comment]
@exhoosh You don't even need a script to select Visual Studio version, as it is embedded in the image name.

**Comment by Dead2** [comment]
@Bagira80 Sorry to have to ask you this, but could you please rebase this on top of develop?
There are no conflicts, but develop contains a fix for the broken CI.

**Comment by DenizThatMenace** [comment]
> @Bagira80 Sorry to have to ask you this, but could you please rebase this on top of develop? There are no conflicts, but develop contains a fix for the broken CI.

Done.

**Pull request merged**
